app/controllers/api/v1/materials_controller.rb

```diff
@@ -44,6 +44,264 @@ def featured
+  #
+  # GET /api/v1/materials/remove_favorite
+  #
+  def remove_favorite
+
+    status  = 200
+    message = "Removing favorite..."
+ 
+    if current_user && !current_user.anonymous?
+      favorite_id = params[:favorite_id]
+
+      if favorite_id
+        favorite = nil
+
+        begin
+          favorite = Favorite.find(favorite_id)
+        rescue ActiveRecord::RecordNotFound => rnf
+          status = 400
+          message = "RecordNotFound Favorite #{favorite_id} does not exist."
+        end
+
+        if favorite
+
+          if favorite.user == current_user
+
+            favorite.destroy
+            message = "Favorite #{favorite_id} removed."
+
+          else
+            status = 400
+            message = "Cannot delete favorite not owned by current user."
+          end
+
+        else
+          status = 400
+          message = "Favorite #{favorite_id} does not exist."
+        end
+
+      else
+        status = 400
+        message = "No favorite id specified."
+      end
+
+    else 
+      status = 400
+      message = "Cannot remove favorite for non-logged in user."
+    end
+
+    render json: {:message => message}, :status => status
+
+  end
+
+  #
+  # Add a favorite to the current user.
+  #
+  # Request params should contain:
+  #
+  # id      The id of the material
+  # type    The type of the material
+  #
+  # GET /api/v1/materials/add_favorite
+  #
+  def add_favorite
+
+    status      = 200
+    message     = "Adding favorite..."
+    favorite_id = -1
+
+    if current_user && !current_user.anonymous?
+    
+      type  = params[:material_type]
+      id    = params[:id]
+
+      if type && id 
+
+        item = nil
+
+        begin
+
+          #
+          # Map of class types supported for favorites.
+          #
+          # Might also consider using "classify" and "constantize" here
+          # rather than a hard coded map of types. But that might require
+          # additional input validation or otherwise constrain how we define
+          # the http parameters.
+          #
+          supported_types = {
+            "external_activity" => ExternalActivity,
+            "interactive"       => Interactive,
+            "investigation"     => Investigation
+          }
+
+          rubyclass = supported_types[type]
+
+          if rubyclass 
+            item = rubyclass.find(id)
+
+            if item
+              #
+              # Unclear if this should first check for the existence of 
+              # the favorite. The unique index should ensure there is only
+              # one favorite per user per item. When attempting to add 
+              # a duplicate, rails logs:
+              # Favorite Exists (1.0ms) SELECT 1 AS one FROM `favorites` ...
+              # and no error is reported. This might be less expensive than
+              # attempting to determine if the favorite exists otherwise.
+              #
+              favorite = Favorite.create(   user: current_user, 
```

Comment:
Yeah I see that ActiveRecord::Relation only got find_or_initialize_by in 4.02: https://apidock.com/rails/v4.2.7/ActiveRecord/Relation/find_or_initialize_by
While first_or_initialize has been there since 3.1.2:
https://apidock.com/rails/v3.2.1/ActiveRecord/Relation/first_or_initialize

It is not clear to me if first_or_initialize will pick up the where conditions. I think it will based on some SO posts. So it might work if you do:
`favorite = current_user.favorites.where(favoritable_id: id, favoritable_type: rubyclass.name).first_or_initialize`
But you might need to do:
`favorite = current_user.favorites.where(favoritable_id: id, favoritable_type: rubyclass.name).first_or_initialize(favoritable: item)`